torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+from __future__ import annotations
+
+import dataclasses
+import functools
+import itertools
+import threading
+import warnings
+import weakref
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+)
+from weakref import ref as WeakRef
+
+import torch.fx
+from torch import Tensor
+
+if torch.has_cuda:
+    from torch._C import _cuda_CUDAAllocator_AllocatorState as AllocatorState
+else:
+    AllocatorState = Any
+
+from torch._dynamo.mutation_guard import GenerationTracker
+
+from torch._inductor.compile_fx import (
+    get_expanded_dims,
+    index_expanded_dims,
+    remove_unaligned_input_idxs,
+    static_input,
+)
+from torch._prims_common import check
+
+from torch.utils import _pytree as pytree
+from . import config
+
+
+@dataclasses.dataclass(frozen=True)
+class GraphID:
+    "Unique counter of a cuda graph recording"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class FunctionID:
+    "Unique counter of a function wrapped in cudagraphify_impl"
+    id: int
+
+
+@dataclasses.dataclass(frozen=True)
+class WrappedFunction:
+    model: Callable
+    static_input_idxs: Sequence[int]
+    id: FunctionID
+
+
+class TreeManagerContainer:
+    """
+    Manages the lifetime of the tree manager. Like `PrivatePool` in cuda caching allocator,
+    the tree and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
+
+    The lifecycle of a tree_manager is:
+    -  Is constructed, no graph, no fns, no tensors
+    -  Tree manager is fetched, resulting in tree manager being allocated
+    -  We generate a bunch of functions, calling add_strong_reference
+    -  These functions die, calling finalize_reference
+    -  When all the functions die, we finalize_tree_manager.
+    -  We look for all the live tensors and add references to THOSE
+    -  We count as tensors die
+    -  All the tensors are dead, we deallocate the tree manager
+    """
+
+    def __init__(self):
+        # This class keeps a strong reference to tree_manager,
+        # but upon all other strong references to the tree_manager will reset it to None.
+        # We need a strong reference so that we can still access its attributes upon cleanup.
+        self.tree_manager: Optional[CUDAGraphTreeManager] = None
+
+        # Number of outstanding references to the current tree manager
+        self.live_cudagraphify_fns = 0
+
+        # Following two objects are only set in the case that Tensor outputs outlive
+        # the cudagraphify_fns. Reference to the Graph is needed to keep the private pool from
+        # deallocation.
+        self.live_tensors_count = 0
+        self.graph: Optional[torch.cuda.CUDAGraph] = None
```

Comment:
Not sure what this is

Reply:
"Reference to the Graph is needed to keep the private pool from deallocation." 

All the cudagraphify_fns are dead, but we keep graph around to prevent deallocation. i simplify this in a later PR.. :'(